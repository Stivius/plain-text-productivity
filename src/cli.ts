#! /usr/bin/env node

import main from "./index";

main().catch((err) => console.log(err, "Error while running"));
