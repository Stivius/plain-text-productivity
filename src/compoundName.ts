const SEPARATOR = "/";

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundName {
  parent: CompoundName;
  name: string;

  constructor(value: string) {
    const splitted = value.split(SEPARATOR);
    if (splitted.length > 0) {
      const name = splitted.pop();
      if (name.length === 0) {
        throw Error("subpart cannot be empty");
      }
      if (splitted.length > 0) {
        this.parent = new CompoundName(splitted.join(SEPARATOR));
      }
      this.name = name;
    } else {
      this.name = value;
    }
  }

  toString(): string {
    if (this.parent != null) {
      return `${this.parent.toString()}/${this.name}`;
    }
    return this.name;
  }

  isEqual(other: CompoundName): boolean {
    return this.toString() === other.toString();
  }
}

export const searchForCompoundProject = (searchedValue: CompoundName) => {
  return (value: CompoundName) => value.isEqual(searchedValue);
};
