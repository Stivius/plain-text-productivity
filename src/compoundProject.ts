const SEPARATOR = '/'

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundProject {
    parent: CompoundProject;
    name: string;

    constructor(value: string) {
        const splitted = value.split(SEPARATOR);
        if (splitted.length > 0) {
            const name = splitted.pop();
            if (name.length === 0) {
                throw Error('subpart cannot be empty');
            }
            if (splitted.length > 0) {
                this.parent = new CompoundProject(splitted.join(SEPARATOR));
            }
            this.name = name;
        } else {
            this.name = value;
        }
    }

    toString(): string {
        if (this.parent != null) {
            return `${this.parent.toString()}/${this.name}`
        }
        return this.name;
    }

    isEqual(other: CompoundProject): boolean {
        return this.toString() === other.toString();
    }
}
