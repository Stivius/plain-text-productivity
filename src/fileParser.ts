import fs from 'fs';
import moment from 'moment';
import {
    FileData,
    Record,
    Mark,
    EstimatedProject,
    DATE_REGEX,
    PROJECT_NAME_REGEX,
    METADATA_PROJECT_NAME_REGEX
} from "./interfaces";

enum ParsingState {
    Init,
    MetadataStarted,
    MetadataActiveProjects,
    ArchivedProjects,
    Blank,
    Record,
}

export class FileParser {
    private projects: string[] = [];
    private records: Record[] = [];
    private currentRecord: Record = undefined;

    constructor(private filename: string) { }

    parse(): FileData {
        const data = fs.readFileSync(this.filename);
        const lines = data.toString().split('\n');

        const METADATA_SEPARATOR = '---';

        let state: ParsingState = ParsingState.Init;
        let metadataLines = 0;

        for (const line of lines) {
            const trimmedLine = line.trim();
            ++metadataLines;
            // TODO: state machine
            if (trimmedLine === METADATA_SEPARATOR) {
                if (state === ParsingState.Init) {
                    state = ParsingState.MetadataStarted;
                } else {
                    state = ParsingState.Blank;
                }
            }
            else if (trimmedLine.indexOf('Projects:') !== -1) {
                if (state === ParsingState.MetadataStarted) {
                    state = ParsingState.MetadataActiveProjects;
                }
            }
            else if (state === ParsingState.MetadataActiveProjects) {
                this.projects.push(this.parseMetadataProject(trimmedLine));
            }
            else if (trimmedLine.length === 0 && state !== ParsingState.Init) {
                state = ParsingState.Blank;
                if (this.currentRecord !== undefined) {
                    this.records.push(this.currentRecord);
                    this.currentRecord = undefined;
                }
            }
            else if (trimmedLine.match(DATE_REGEX) && state === ParsingState.Blank) {
                state = ParsingState.Record;
                this.currentRecord = this.createRecord(trimmedLine);
            }
            else if (state === ParsingState.Record) {
                this.currentRecord.projects.push(this.parseProject(trimmedLine));
            }
        }
        return { metadata: { projects: this.projects }, records: this.records };
    }

    private createRecord(line: string) {
        if (!moment(line, 'YYYY-MM-DD').isValid()) {
            throw Error('invalid date')
        }
        return {
            day: new Date(line),
            projects: [],
        }
    }

    private parseMetadataProject(line: string) {
        if (!line.match(METADATA_PROJECT_NAME_REGEX)) {
            throw Error('invalid metadata project');
        }
        const parsedProject = line.replace('-', '').trim();
        if (this.projects.indexOf(parsedProject) !== -1) {
            throw Error('duplicated metadata project');
        }
        return parsedProject;
    }

    private parseProject(line: string): EstimatedProject {
        const splitted = line.split(':');
        const [ name, mark ] = splitted;
        if (splitted.length !== 2 || this.currentRecord === undefined || !name.match(PROJECT_NAME_REGEX)) {
            throw new Error('invalid record')
        }
        if (this.currentRecord.projects.find((value) => value.name === name) !== undefined) {
            throw new Error('duplicated project record')
        }
        return {
            name: splitted[0],
            mark: this.parseMark(mark),
        };
    }

    private parseMark(value: string): Mark {
        // TODO: mark regex validate
        const trimmedValue = value.trim();
        if (trimmedValue === '-') {
            return undefined;
        }
        const mark = parseInt(value);
        if (Number.isNaN(mark) || mark < 1 || mark > 5) {
            throw new Error('invalid mark')
        }
        return mark as Mark;
    }
}
