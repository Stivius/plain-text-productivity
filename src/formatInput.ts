import chalk from "chalk";
import columnify from "columnify";
import dateFormat from "dateformat";
import prompt from "prompt";

function formatDate(date: Date) {
  return dateFormat(date, "yyyy-mm-dd");
}

export function formatInputForFile(date: Date, data: prompt.Properties) {
  return `${formatDate(date)}\n`
    .concat(
      Object.keys(data)
        .map((k) => {
          return `${k}:${data[k]}`;
        })
        .join("\n")
    )
    .concat("\n\n");
}

export function formatInputForConsole(
  date: Date,
  data: prompt.Properties
): string {
  const colorizeMark = (value: string) => {
    switch (value) {
      case "1":
        return chalk.rgb(150, 0, 0).bold(value);
      case "2":
        return chalk.rgb(255, 125, 125).bold(value);
      case "3":
        return chalk.rgb(255, 255, 75).bold(value);
      case "4":
        return chalk.rgb(125, 255, 125).bold(value);
      case "5":
        return chalk.rgb(0, 180, 0).bold(value);
    }
    return chalk.white(value);
  };

  return `${chalk.blue(formatDate(date))}\n`.concat(
    columnify(
      Object.keys(data).map((k) => ({
        project: chalk.yellow(k),
        mark: colorizeMark(data[k] as string),
      }))
    )
  );
}
