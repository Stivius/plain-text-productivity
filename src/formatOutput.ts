import chalk from 'chalk';
import columnify from 'columnify'

export function formatDataForFile(data) {
	return Object.keys(data).map((k) => {
		if (k !== 'Date') {
			return `${k}:${data[k]}`
		} else {
			return `${data[k]}`;
		}
	}).join('\n').concat('\n');
}

export function outputDataToConsole(data) {
	const colorizeMark = (value: string) => {
		switch (value) {
			case '1': return chalk.rgb(150, 0, 0).bold(value); 
			case '2': return chalk.rgb(255, 125, 125).bold(value); 
			case '3': return chalk.rgb(255, 255, 75).bold(value); 
			case '4': return chalk.rgb(125, 255, 125).bold(value); 
			case '5': return chalk.rgb(0, 180, 0,).bold(value); 
		}
		return chalk.white(value);
	};

	const DATE_KEY = 'Date';
	if (data.hasOwnProperty(DATE_KEY)) {
		console.log(`${chalk.blue(data[DATE_KEY])}`);
		delete data[DATE_KEY];
	}

	console.log(columnify(Object.keys(data).map((k) => ({
		project: chalk.yellow(k),
		mark: colorizeMark(data[k] as string),
	}))));
}

