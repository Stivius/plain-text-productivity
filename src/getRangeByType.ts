import dateFormat from "dateformat";

enum Aggregation {
    Day,
    Week,
    Month,
    Year
};

const dayInThePast = (daysToSubtract: number = 0) => {
    const date = new Date();
    date.setDate(date.getDate() + daysToSubtract);
    return date;
};

const beginingOf = (aggregation: Aggregation) => {
    const date = new Date();
    switch (aggregation) {
        case Aggregation.Day: return date;
        case Aggregation.Week: {
            const day = date.getDay();
            const diff = date.getDate() - day + (day == 0 ? -6 : 1);
            return new Date(date.setDate(diff));
        }
        case Aggregation.Month: return new Date(date.getFullYear(), date.getMonth(), 1);
        case Aggregation.Year: return new Date(date.getFullYear(), 0, 1);
    }
};

// TODO: change number to enum
export function getRangeByType(choice: number) {
    if (isNaN(choice) || choice < 1 || choice > 10) {
        throw new Error(`${choice} is not valid range type`);
    }

    const choiceMappings = {
        1: beginingOf(Aggregation.Day),
        2: beginingOf(Aggregation.Week),
        3: beginingOf(Aggregation.Month),
        4: beginingOf(Aggregation.Year),
        5: dayInThePast(-7),
        6: dayInThePast(-14),
        7: dayInThePast(-30),
        8: dayInThePast(-90),
        9: dayInThePast(-180),
        10: dayInThePast(-365),
    };

    return { 
        from: new Date(dateFormat(choiceMappings[choice], "yyyy-mm-dd")),
        to: new Date(dateFormat(new Date(), "yyyy-mm-dd")),
        choice: choice,
    };
}

