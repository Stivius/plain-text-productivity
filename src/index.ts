import commandLineArgs from 'command-line-args';
import fs from 'fs';
import dateFormat from 'dateformat'
import {readRecordFromConsole} from './readRecordFromInput';
import chalk from 'chalk';
import {formatDataForFile, outputDataToConsole} from './formatOutput';
import {EstimatedProject, FileData, FILE_NAME, ProjectRecord} from './interfaces';
import {readRangeChoiceFromConsole} from './readRangeChoiceFromInput';
import {getRangeByType} from './getRangeByType';
import {FileParser} from './fileParser';
import {generateReport} from './generateReport';
import {outputReport} from './outputReport';

const commandsDefinitions = [
    { name: 'command', defaultOption: true },
]

const commands = commandLineArgs(commandsDefinitions, { stopAtFirstUnknown: true });

async function main() {
    try {
        const parser = new FileParser(FILE_NAME);
        const data: FileData = parser.parse();

        switch (commands.command) {
            // TODO: refactor output to console/file
            case 'add': {
                const enteredData = await readRecordFromConsole(data.metadata.activeProjects);
                const date = dateFormat(new Date(), "yyyy-mm-dd");
                console.log(chalk.blue(date));
                fs.appendFileSync(FILE_NAME, date.concat('\n', formatDataForFile(enteredData), '\n'));
                outputDataToConsole(enteredData);
                break;
            }

            case 'addp': {
                const enteredData = await readRecordFromConsole(data.metadata.activeProjects, true);
                fs.appendFileSync(FILE_NAME, formatDataForFile(enteredData).concat('\n'));
                outputDataToConsole(enteredData);
                break;
            }

            case 'lsp':
                const projects = data.records.map((r: ProjectRecord) => r.projects.map((p: EstimatedProject) => p.name));
            const uniqueProjects = Array.from(new Set(projects.flat()));
            const achivedProjects = uniqueProjects.filter((p) => !data.metadata.activeProjects.includes(p));
            console.log('All projects:');
            console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
            console.log(chalk.gray(achivedProjects.join('\n')));
            break;

            case 'lspa':
                console.log('Active projects:');
            console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
            break;

            case 'report': {
                const optionDefinitions = [
                    { name: 'absolute', type: Boolean, defaultValue: false },
                    { name: 'range-type', type: Number },
                    { name: 'from-date', type: String },
                    { name: 'to-date', type: String },
                ]
                const options = commandLineArgs(optionDefinitions, { argv: commands._unknown || [] });
                const getRange = (options: commandLineArgs.CommandLineOptions) => {
                    const rangeType = options['range-type'];
                    if (rangeType === undefined) {
                        return readRangeChoiceFromConsole();
                    } else {
                        // TODO: can be entered as string and converted to number
                        if (rangeType === 0) {
                            const from = new Date(options['from-date']);
                            const to = new Date(options['to-date']);
                            if (!from || !to) {
                                throw new Error('Date range is not specified or invalid');
                            }
                            return { from, to, choice: 0 };
                        }
                        return getRangeByType(rangeType);
                    }
                };

                const { from, to, choice } = await getRange(options);
                const productivityByProject = generateReport(data.records, from, to, options.absolute);
                outputReport(productivityByProject, choice, from, to, options.absolute);
                break;
            }
            default:
                console.error('Unknown command');
            break;
        }
    } catch (error) {
        console.error(error);
    }
}

main().catch((err) => console.log(err, 'Error while running'))
