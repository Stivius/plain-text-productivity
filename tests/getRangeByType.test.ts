import { getRangeByType } from "../src/getRangeByType";

describe("getRangeByType", () => {
  beforeAll(() => {
    jest.useFakeTimers("modern");
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test.each([
    new Date(2022, 6, 1),
    new Date(2022, 6, 5),
    new Date(2022, 6, 30),
  ])("check daily range", (date: Date) => {
    jest.setSystemTime(date);
    expect(getRangeByType(1)).toMatchSnapshot();
  });

  test.each([
    new Date(2022, 4, 30),
    new Date(2022, 4, 31),
    new Date(2022, 5, 1),
    new Date(2022, 5, 2),
    new Date(2022, 5, 3),
    new Date(2022, 5, 4),
    new Date(2022, 5, 5),
    new Date(2022, 5, 6),
  ])("check weekly range", (date: Date) => {
    jest.setSystemTime(date);
    expect(getRangeByType(2)).toMatchSnapshot();
  });

  test.each([
    new Date(2022, 6, 1),
    new Date(2022, 6, 5),
    new Date(2022, 6, 30),
  ])("check monthly range", (date: Date) => {
    jest.setSystemTime(date);
    expect(getRangeByType(3)).toMatchSnapshot();
  });

  test.each([
    new Date(2022, 0, 1),
    new Date(2022, 6, 15),
    new Date(2022, 11, 31),
  ])("check yearly range", (date: Date) => {
    jest.setSystemTime(date);
    expect(getRangeByType(4)).toMatchSnapshot();
  });

  test.each([5, 6, 7, 8, 9, 10])("check n-days range", (type: number) => {
    jest.setSystemTime(new Date(2022, 6, 1));
    expect(getRangeByType(type)).toMatchSnapshot();
  });

  test.each([
    "dummy",
    0,
    -1,
    11,
    Number.NEGATIVE_INFINITY,
    Number.POSITIVE_INFINITY,
    Number.MIN_SAFE_INTEGER,
    Number.MAX_SAFE_INTEGER,
  ])("check wrong type", (type: number) => {
    jest.setSystemTime(new Date(2022, 6, 1));
    expect(() => getRangeByType(type)).toThrowError();
  });
});
