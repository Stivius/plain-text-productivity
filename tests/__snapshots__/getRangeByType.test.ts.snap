// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRangeByType check daily range 1`] = `
Object {
  "choice": 1,
  "from": 2022-07-01T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check daily range 2`] = `
Object {
  "choice": 1,
  "from": 2022-07-05T00:00:00.000Z,
  "to": 2022-07-05T00:00:00.000Z,
}
`;

exports[`getRangeByType check daily range 3`] = `
Object {
  "choice": 1,
  "from": 2022-07-30T00:00:00.000Z,
  "to": 2022-07-30T00:00:00.000Z,
}
`;

exports[`getRangeByType check monthly range 1`] = `
Object {
  "choice": 3,
  "from": 2022-07-01T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check monthly range 2`] = `
Object {
  "choice": 3,
  "from": 2022-07-01T00:00:00.000Z,
  "to": 2022-07-05T00:00:00.000Z,
}
`;

exports[`getRangeByType check monthly range 3`] = `
Object {
  "choice": 3,
  "from": 2022-07-01T00:00:00.000Z,
  "to": 2022-07-30T00:00:00.000Z,
}
`;

exports[`getRangeByType check n-days range 1`] = `
Object {
  "choice": 5,
  "from": 2022-06-24T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check n-days range 2`] = `
Object {
  "choice": 6,
  "from": 2022-06-17T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check n-days range 3`] = `
Object {
  "choice": 7,
  "from": 2022-06-01T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check n-days range 4`] = `
Object {
  "choice": 8,
  "from": 2022-04-02T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check n-days range 5`] = `
Object {
  "choice": 9,
  "from": 2022-01-02T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check n-days range 6`] = `
Object {
  "choice": 10,
  "from": 2021-07-01T00:00:00.000Z,
  "to": 2022-07-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 1`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-05-30T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 2`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-05-31T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 3`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-06-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 4`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-06-02T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 5`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-06-03T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 6`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-06-04T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 7`] = `
Object {
  "choice": 2,
  "from": 2022-05-30T00:00:00.000Z,
  "to": 2022-06-05T00:00:00.000Z,
}
`;

exports[`getRangeByType check weekly range 8`] = `
Object {
  "choice": 2,
  "from": 2022-06-06T00:00:00.000Z,
  "to": 2022-06-06T00:00:00.000Z,
}
`;

exports[`getRangeByType check yearly range 1`] = `
Object {
  "choice": 4,
  "from": 2022-01-01T00:00:00.000Z,
  "to": 2022-01-01T00:00:00.000Z,
}
`;

exports[`getRangeByType check yearly range 2`] = `
Object {
  "choice": 4,
  "from": 2022-01-01T00:00:00.000Z,
  "to": 2022-07-15T00:00:00.000Z,
}
`;

exports[`getRangeByType check yearly range 3`] = `
Object {
  "choice": 4,
  "from": 2022-01-01T00:00:00.000Z,
  "to": 2022-12-31T00:00:00.000Z,
}
`;
